### ISSUE TITLE
Restricted prototype of object suddenly changes to something else

### ISSUE THREAD

**Issue by KurtGokhan** [microsoft/ClearScript#282]
I tried to use restricted prototype workaround we discussed in #269 but there was a problem. The restricted prototype works good initially. But after a while, the prototype of the object magically changes to another completely unrelated host object. It usually happens after about 3-5 seconds of inactivity, or completely randomly even when there is activity.

I suspect it has something to do with either GC or caching. I tried to prevent the restricted prototype from being GCed but the problem persisted. I believe the problem is in the unmanaged code, either in ClearScript.V8 or V8 itself.

My code is quite complex so it may not be a good example. I have been trying to reproduce this error in an isolated context but I failed to do so. I am opening this issue in case you may have an idea what is happening.

**Comment by ClearScriptLib** [comment]
Hi @KurtGokhan,

Have you updated to ClearScript 7.1.5? Earlier releases had a caching issue that affected your specific scenario involving `OnExposedToScriptCode`.

Also, it might be helpful to find out _what_ the prototype is unexpectedly changing to. Is it a host object? If so, what is its type? Etc.

Thanks!

**Comment by KurtGokhan** [comment]
@ClearScriptLib  I am trying to reproduce it like this:

```
        [TestMethod, TestCategory("BugFix")]
        public async Task BugFix_RestrictedPrototypeMagicallyChanging()
        {
            var test = new RestrictedPrototypeObject(true);
            engine.Script.Assert = HostType.Wrap(typeof(Assert));
            engine.Script.test = test;
            Assert.AreEqual("bar", engine.Evaluate("test.Foo"));

            Func<dynamic, double, Task> setTimeout = async (dynamic callback, double timeout) =>
            {
                await Task.Delay((int) timeout);
                callback();
            };

            engine.Script.setTimeout = setTimeout;

            for (int i = 0; i < 6; i++)
            {
                var task = engine.Evaluate(@"setTimeout(() => 
                    Assert.AreEqual('Microsoft.ClearScript.Test.BugFixTest+RestrictedPrototypeObject', test.__proto__.ToString()), 5000);") as Task;
                await task;
            }
        }
```

The issue does not happen here but it happens in my program where I basically do the same thing. In my program there are other objects, a lot of async stuff (single threaded) and a large source code so that may be causing some complex flow of events. Although nothing should be affecting this object directly.

What happens is, type of `__proto__` is initially `Microsoft.ClearScript.Test.BugFixTest+RestrictedPrototypeObject`, which is correct. Then after a while, it becomes `Microsoft.ClearScript.V8.V8ScriptItem`. I checked it with debugger and I think it has the type `HostInvocable`.

**Comment by ClearScriptLib** [comment]
@KurtGokhan Are you by any chance exposing your object in more than one `V8ScriptEngine` instance?

Also, could you post your implementation of `OnExposedToScriptCode`, or at least a portion of it that you can share?

**Comment by KurtGokhan** [comment]
The entire class with `OnExposedToScriptCode` is [here](https://github.com/ReactUnity/core/blob/main/Runtime/Helpers/WatchableRecord.cs#L249).

I wrote it to handle multiple engines but it is actually using 1 engine instance.

Btw, I am using 7.1.5.

This error is so weird that the prototype can disappear even when debugging. See:

```
    console.log(typeof test.Set); // --> function
    debugger;
    console.log(typeof test.Set); // --> undefined
```

When I hit the `debugger`, `test.Set` is undefined. But in the previous line it was not undefined.

**Comment by ClearScriptLib** [comment]
Hi @KurtGokhan,

Hmm, that _is_ weird. You might be able to use a [JavaScript proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) to capture the prototype reassignment:

``` JavaScript
test = new Proxy(test, {
    setPrototypeOf(target, proto) {
        debugger;
        return Reflect.setPrototypeOf(target, proto);
    }
});
```

Beyond that, it's difficult to diagnose the issue without a test case.

Good luck!

**Comment by KurtGokhan** [comment]
I could not reproduce it in isolation yet, I will update when I can.

I think nothing is modifying the prototype other than the initial code in `OnExposedToScriptCode`. I think the object at the place pointer is pointing to changes somehow. I don't know how that is possible though.

**Comment by KurtGokhan** [comment]
Created a test case reproducing the issue in #284. But no solution yet.

**Comment by KurtGokhan** [comment]
The issue was addressed in #284. Closing.

Summary: This is not a bug, proxy object is garbage-collected in V8.
